sinks/grpsink/grpc_stream.go

```diff
@@ -0,0 +1,273 @@
+	gs.traceClient = cl
+
+	var err error
+	gs.stream, err = gs.sinkClient.SendSpans(context.TODO())
+	if err != nil {
+		gs.log.WithFields(logrus.Fields{
+			"name":          gs.name,
+			"target":        gs.target,
+			"chanstate":     gs.grpcConn.GetState().String(),
+			logrus.ErrorKey: err,
+		}).Error("Failed to set up a stream over gRPC channel to sink target")
+		return err
+	}
+
+	go gs.maintainStream()
+	return nil
+}
+
+// maintainStream is intended to be run in a background goroutine. It is kicked
+// off by Start(), and is only kept as a standalone method for testing purposes.
+//
+// This method runs an endless loop that reacts to state changes in the
+// underlying channel by automatically attempting to re-establish the stream
+// connection when it moves back into the 'READY' state. See
+// https://github.com/grpc/grpc/blob/master/doc/connectivity-semantics-and-api.mdA
+// for details about gRPC's connectivity state machine.
+func (gs *GRPCStreamingSpanSink) maintainStream() {
+	for {
+		// This call will block on a channel receive until the gRPC connection
+		// becomes unhealthy. Then, spring into action!
+		state := gs.grpcConn.GetState()
+		switch state {
+		case connectivity.Idle:
+			// gRPC considers an open stream to be an active RPC, and channels
+			// normally only go idle if there are no active RPC within the
+			// channel's timeout window. It should be nigh-impossible for that
+			// to happen here, so we record entering the Idle state as an error.
+			gs.log.WithFields(logrus.Fields{
+				"name":          gs.name,
+				"target":        gs.target,
+				"chanstate":     gs.grpcConn.GetState().String(),
+				logrus.ErrorKey: fmt.Errorf("gRPC sink became idle; should be nearly impossible"),
+			}).Error("gRPC channel transitioned to idle")
+
+			// With the error sent, now fall through to recreating the stream
+			// as if the channel were ready, as stream creation should force
+			// the channel back out of idle. This isn't ideal - it'll end up
+			// recreating the stream again after re-reaching READY - but it's
+			// not a huge problem, as this case is essentially unreachable.
+			fallthrough
+		case connectivity.Ready:
+			gs.streamMut.Lock()
+			// CAS inside the mutex so that we can't ever get a double-send of the
+			// "first time channel degradation" log error.
+			atomic.CompareAndSwapUint32(&gs.bad, 1, 0)
+			stream, err := gs.sinkClient.SendSpans(context.TODO())
+			if err != nil {
+				// This is a weird case and probably shouldn't be reachable, but
+				// if stream setup fails after recovery, then just wait for a
+				// second and try again.
+				gs.log.WithFields(logrus.Fields{
+					"name":          gs.name,
+					"target":        gs.target,
+					"chanstate":     gs.grpcConn.GetState().String(),
+					logrus.ErrorKey: err,
+				}).Error("Failed to set up a new stream after gRPC channel recovered")
+				time.Sleep(1 * time.Second)
+				gs.streamMut.Unlock()
```

Comment:
it it necessary to hold the lock while sleeping there?

Reply:
Also, wonder if instead of sleeping this can use `WaitForStateChange` with a longer timeout instead.

Reply:
Ah no, those can be switched around - unlock before sleep.

We don't want another `WaitForStateChange()` here, though, because we shouldn't expect to see one. The state change is in the underlying channel state, not the stream - and being here means that the channel is already back in READY state. If we were to wait again, it might be for a long time - until the next network event disrupts the connection.